### ISSUE TITLE
Cannot use a Theory accepting a generic array

### ISSUE THREAD

**Issue by mellinoe** [xunit/xunit#805]
Consider this test case:
````CSharp
[Theory]
[InlineData(new int[] { 1, 2, 3 })]
public static void Test<T>(T[] array)
{
    foreach (var i in array)
    {
        Console.WriteLine(i);
    }
}
````
An exception occurs when the test case is run. Here is the stack trace:
````
ClassLibrary1.Class1.Test<Object>(array: [1, 2, 3]) [FAIL]
      System.ArgumentException : Object of type 'System.Int32[]' cannot be converted to type 'System.Object[]'.
      Stack Trace:
           at System.RuntimeType.TryChangeType(Object value, Binder binder, CultureInfo culture, Boolean needsSpecialCast)
           at System.RuntimeType.CheckValue(Object value, Binder binder, CultureInfo culture, BindingFlags invokeAttr)
           at System.Reflection.MethodBase.CheckArguments(Object[] parameters, Binder binder, BindingFlags invokeAttr, CultureInfo culture, Signature sig)
           at System.Reflection.RuntimeMethodInfo.InvokeArgumentsCheck(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
           at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
        --- End of stack trace from previous location where exception was thrown ---
           at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
           at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
        --- End of stack trace from previous location where exception was thrown ---
           at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
           at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
````

Based on the above, it seems that the wrong generic method is being created for the test method, i.e. `Test<Object>()` rather than `Test<Int32>()`.

Another thing to note: if a generic constrant is applied to the test method, i.e.:

````
public static void Test<T>(T[] array) where T : struct
````

this also fails, with a different exception: 

````
ClassLibrary1.Class1.Test [FAIL]
      System.ArgumentException : GenericArguments[0], 'System.Object', on 'Void Test[T](T[])' violates the constraint of type 'T'.
      ---- System.Security.VerificationException : Method ClassLibrary1.Class1.Test: type argument 'System.Object' violates the constraint of type parameter 'T'.
      Stack Trace:
           at System.RuntimeType.ValidateGenericArguments(MemberInfo definition, RuntimeType[] genericArguments, Exception e)
           at System.Reflection.RuntimeMethodInfo.MakeGenericMethod(Type[] methodInstantiation)
        ----- Inner Stack Trace -----
           at System.RuntimeMethodHandle.GetStubIfNeeded(RuntimeMethodHandleInternal method, RuntimeType declaringType, RuntimeType[] methodInstantiation)
           at System.Reflection.RuntimeMethodInfo.MakeGenericMethod(Type[] methodInstantiation
````
I believe this exception occurs when the test case is being generated. Presumably calling `MakeGenericMethod` with T=System.Object fails because it does not meet the constraints.

As another side note, the test case with the generic constraint cause the VS test runner to "run continuously" trying to discover the tests and seems to get stuck in a loop.

**Comment by hughbe** [comment]
@mellinoe 
Please take a look at #965 - it adds support for this and a whole lot of other situations!

**Comment by bradwilson** [comment]
Fixed by #965
